scripts/generate_source.py

```diff
@@ -0,0 +1,111 @@
+#!/usr/bin/env python3
+# Copyright (c) 2019 The Khronos Group Inc.
+# Copyright (c) 2019 Valve Corporation
+# Copyright (c) 2019 LunarG, Inc.
+# Copyright (c) 2019 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Author: Mike Schuchardt <mikes@lunarg.com>
+
+import argparse
+import common_codegen
+import filecmp
+import os
+import shutil
+import subprocess
+import sys
+import tempfile
+
+# files to exclude from --verify check
+verify_exclude = ['.clang-format']
+
+def main(argv):
+    parser = argparse.ArgumentParser(description='Generate source code for this repository')
+    parser.add_argument('registry', metavar='REGISTRY_PATH', help='path to the Vulkan-Headers registry directory')
+    group = parser.add_mutually_exclusive_group()
+    group.add_argument('-i', '--incremental', action='store_true', help='only update repo files that change')
+    group.add_argument('-v', '--verify', action='store_true', help='verify repo files match generator output')
+    args = parser.parse_args(argv)
+
+    gen_cmds = [[common_codegen.repo_relative('scripts/kvt_genvk.py'),
+                 '-registry', os.path.abspath(os.path.join(args.registry,  'vk.xml')),
+                 '-quiet',
+                 filename] for filename in ['vk_typemap_helper.h',
+                                            'mock_icd.h',
+                                            'mock_icd.cpp']]
+
+    repo_dir = common_codegen.repo_relative('icd/generated')
+
+    # get directory where generators will run
+    if args.verify or args.incremental:
+        # generate in temp directory so we can compare or copy later
+        temp_obj = tempfile.TemporaryDirectory(prefix='VulkanLoader_generated_source_')
+        temp_dir = temp_obj.name
+        gen_dir = temp_dir
+    else:
+        # generate directly in the repo
+        gen_dir = repo_dir
+
+    # run each code generator
+    for cmd in gen_cmds:
+        print(' '.join(cmd))
+        try:
+            subprocess.check_call([sys.executable] + cmd,
+                                  # ignore generator output, vk_validation_stats.py is especially noisy
+                                  stdout=subprocess.DEVNULL,
+                                  cwd=gen_dir)
+        except Exception as e:
+            print('ERROR:', str(e))
```

Comment:
If an exception does occur, what's the expected developer response?

I think the developer will have to examine the script, figure out by trial and error which `check_call()` execution failed, look at the output, and then figure out what went wrong and fix it.

It would be easier if the `check_call()` saved its stdout and stderr, and printed the command line, working directory, any interesting environment variables, return code, and stdout and stderr in the case of error.  Then the developer wouldn't have to dissect the script and do experiments to figure out what happened.

Reply:
I don't think the developer will have to do very much archaeology to find the error location.  The command is printed to stdout before executing, and stderr is unperturbed.  Here's what happens if I put a whoops in mock_icd_generator.py

```
$ make VulkanTools_generated_source   
/home/mike/Projects/Vulkan-Tools/scripts/kvt_genvk.py -registry /home/mike/Projects/Vulkan-Headers/build/install/share/vulkan/registry/vk.xml -quiet vk_typemap_helper.h
/home/mike/Projects/Vulkan-Tools/scripts/kvt_genvk.py -registry /home/mike/Projects/Vulkan-Headers/build/install/share/vulkan/registry/vk.xml -quiet mock_icd.h
Traceback (most recent call last):
  File "/home/mike/Projects/Vulkan-Tools/scripts/kvt_genvk.py", line 356, in <module>
    genTarget(args)
  File "/home/mike/Projects/Vulkan-Tools/scripts/kvt_genvk.py", line 225, in genTarget
    reg.apiGen(options)
  File "/home/mike/Projects/Vulkan-Headers/build/install/share/vulkan/registry/reg.py", line 1061, in apiGen
    self.gen.beginFile(self.genOpts)
  File "/home/mike/Projects/Vulkan-Tools/scripts/mock_icd_generator.py", line 1078, in beginFile
    raise Exception('whoops')
Exception: whoops
ERROR: Command '['/usr/bin/python3', '/home/mike/Projects/Vulkan-Tools/scripts/kvt_genvk.py', '-registry', '/home/mike/Projects/Vulkan-Headers/build/install/share/vulkan/registry/vk.xml', '-quiet', 'mock_icd.h']' returned non-zero exit status 1.
make[3]: *** [CMakeFiles/VulkanTools_generated_source.dir/build.make:57: CMakeFiles/VulkanTools_generated_source] Error 1
make[2]: *** [CMakeFiles/Makefile2:73: CMakeFiles/VulkanTools_generated_source.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:80: CMakeFiles/VulkanTools_generated_source.dir/rule] Error 2
make: *** [Makefile:164: VulkanTools_generated_source] Error 2
```